File: src-tauri/src/session.rs
//! Session lifecycle management.
//!
//! Each session corresponds to one `alayacore --rawio` subprocess.
//! This module manages creating, resuming, closing, and forking sessions.

use crate::alayacore;
use crate::event::StatusEvent;
use crate::ModelCache;

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::atomic::AtomicBool;
use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio::sync::Mutex;

/// Internal handle to a running alayacore session.
pub struct SessionHandle {
    /// stdin pipe; `None` once closed (graceful close / EOF sent).
    /// The Option lets close_session drop the pipe even while other
    /// Arcs exist (in-flight prompt sends lock the same mutex).
    pub stdin: Arc<Mutex<Option<std::process::ChildStdin>>>,
    pub connected: Arc<AtomicBool>,
    /// Pending command call IDs → command names (CI sent, CO not yet received).
    /// Used by the stdout reader to attach the command name to CO frames
    /// (CO carries only the call ID; the name comes from the CI we sent).
    /// Bounded — see PendingCommands (M5/D6).
    pub pending_commands: Arc<PendingCommands>,
    /// The child process — kept for explicit cleanup on close.
    /// Uses std::sync::Mutex so Drop can access it (sync context).
    pub child: Arc<std::sync::Mutex<Option<std::process::Child>>>,
    /// Path to the session's directory (~/.alayaface/sessions/<uuid>/).
    pub session_dir: PathBuf,
    /// Client identity that created/resumed this session (empty =
    /// legacy). close_all_sessions reclaims only the caller's own
    /// sessions, so one client's page load never kills another client's
    /// live sessions (the Go backend is reachable from several clients).
    pub owner: String,
}

/// Bounded pending-command registry: call ID → command name (CI sent,
/// CO not yet received). If a CO reply never arrives (protocol anomaly
/// / killed core), an entry must not grow the map forever — insert
/// evicts the OLDEST entries beyond the cap (FIFO), so the newest
/// calls — the ones most likely to still get a reply — survive.
/// Mirrors the Go internal/session/pendingCmds guard (M5/D6).
/// `pub` because it appears in the signature of the public
/// `spawn_stdout_reader` and the public `SessionHandle::pending_commands`
/// field; its internals stay private.
pub struct PendingCommands {
    inner: tokio::sync::Mutex<PendingCommandsInner>,
}

struct PendingCommandsInner {
    map: std::collections::HashMap<String, String>,
    order: std::collections::VecDeque<String>,
}

/// Maximum pending CI→CO entries per session.
pub(crate) const MAX_PENDING_COMMANDS: usize = 512;

impl PendingCommands {
    pub(crate) fn new() -> Arc<Self> {
        Arc::new(Self {
            inner: tokio::sync::Mutex::new(PendingCommandsInner {
                map: std::collections::HashMap::new(),
                order: std::collections::VecDeque::new(),
            }),
        })
    }

    /// Record id → name, evicting the oldest entries beyond the cap.
    pub(crate) async fn insert(&self, id: String, name: String) {
        let mut g = self.inner.lock().await;
        let is_new = g.map.insert(id.clone(), name).is_none();
        if is_new {
            g.order.push_back(id);
        }
        while g.order.len() > MAX_PENDING_COMMANDS {
            if let Some(old) = g.order.pop_front() {
                g.map.remove(&old);
            }
        }
    }

    /// Remove and return the name for id.
    pub(crate) async fn remove(&self, id: &str) -> Option<String> {
        let mut g = self.inner.lock().await;
        let v = g.map.remove(id);
        if v.is_some() {
            g.order.retain(|x| x != id);
        }
        v
    }

    /// Remove and return the name for id, from a non-async context
    /// (the stdout reader thread).
    pub(crate) fn blocking_remove(&self, id: &str) -> Option<String> {
        let mut g = self.inner.blocking_lock();
        let v = g.map.remove(id);
        if v.is_some() {
            g.order.retain(|x| x != id);
        }
        v
    }

    #[cfg(test)]
    async fn len(&self) -> usize {
        self.inner.lock().await.map.len()
    }

    #[cfg(test)]
    async fn contains(&self, id: &str) -> bool {
        self.inner.lock().await.map.contains_key(id)
    }
}

impl Drop for SessionHandle {
    fn drop(&mut self) {
        // Best-effort EOF so alayacore can drain + auto-save before the
        // kill below gives it a short grace period (kill_child now waits
        // for a natural exit before SIGKILL).
        if let Ok(mut guard) = self.stdin.try_lock() {
            *guard = None;
        }
        if let Ok(mut guard) = self.child.lock() {
            if let Some(mut child) = guard.take() {
                alayacore::kill_child(&mut child);
            }
        }
    }
}

/// Shared map of session_id → SessionHandle.
pub struct SessionMap(pub Arc<Mutex<HashMap<String, SessionHandle>>>);

/// Configuration for creating a new session.
pub struct SessionConfig<'a> {
    /// Session id — also names the on-disk directory
    /// (~/.alayaface/sessions/<id>/), matching the Go backend.
    pub id: &'a str,
    pub app: &'a AppHandle,
    pub binary: &'a str,
    pub config_path: &'a str,
    pub session_file: &'a str,
    pub session_dir: PathBuf,
    pub sessions: &'a SessionMap,
    pub model_cache: &'a ModelCache,
    pub tool_confirm: &'a str,
    /// Built-in tools: Some(list) → `--builtin-tools=<list>` (Some("") =
    /// NO builtin tools — used by Plan Sessions so the planner cannot
    /// execute tools); None = don't pass the flag = all tools.
    pub builtin_tools: Option<&'a str>,
    pub system_prompt: &'a str,
    /// Initial reasoning level (0|1|2), passed to alayacore as
    /// `--reasoning-level=<n>`.
    pub reasoning_level: i64,
    /// Child process working directory (per-plan isolation; None =
    /// inherit the backend's cwd).
    pub work_dir: Option<String>,
    /// Client identity that created the session (empty = legacy). Used
    /// by close_all_sessions to reclaim only one client's orphaned
    /// sessions.
    pub owner: &'a str,
}

// ─── Factory ──────────────────────────────────────────────────────────

/// Create a new session: spawn alayacore and start background readers.
/// The session id comes from the caller (cfg.id) so the returned id
/// always equals the on-disk directory name — resume_session depends on
/// this (it looks up the dir by the id it is handed).
pub async fn create(cfg: SessionConfig<'_>) -> Result<String, String> {
    let session_id = cfg.id.to_string();

    let proc = alayacore::spawn(
        cfg.binary,
        cfg.config_path,
        cfg.session_file,
        cfg.tool_confirm,
        cfg.builtin_tools,
        cfg.system_prompt,
        cfg.reasoning_level,
        cfg.work_dir.as_deref(),
    )
    .map_err(|e| format!("Failed to start alayacore: {e}"))?;

    let connected = Arc::new(AtomicBool::new(true));
    let stdin = Arc::new(Mutex::new(Some(proc.stdin)));
    let child = Arc::new(std::sync::Mutex::new(Some(proc.child)));
    let pending_commands = PendingCommands::new();

    let handle = SessionHandle {
        stdin: stdin.clone(),
        connected: connected.clone(),
        pending_commands: pending_commands.clone(),
        child: child.clone(),
        session_dir: cfg.session_dir,
        owner: cfg.owner.to_string(),
    };

    cfg.sessions.0.lock().await.insert(session_id.clone(), handle);

    // Emit connected:true BEFORE spawning the stdout reader: a child
    // that dies instantly makes the reader emit connected:false right
    // away — with the old ordering (reader first) the false could land
    // before the true, leaving the client believing a dead session is
    // connected.
    let _ = cfg.app.emit("core-status", StatusEvent {
        session_id: session_id.clone(),
        connected: true,
        message: format!("Connected to alayacore ({})", cfg.binary),
    });

    // Background reader for stdout
    crate::reader::spawn_stdout_reader(
        cfg.app.clone(),
        session_id.clone(),
        proc.stdout,
        connected,
        cfg.model_cache.0.clone(),
        child.clone(),
        pending_commands,
    );

    Ok(session_id)
}

/// Close a session gracefully: ask alayacore to save, send EOF (it
/// drains the active task — auto-saving at task end — then exits), and
/// only SIGKILL after a grace period. See `alayacore::close_child_gracefully`.
pub async fn close(session_id: &str, sessions: &SessionMap) -> Result<(), String> {
    close_with_timeout(session_id, sessions, alayacore::GRACEFUL_CLOSE_TIMEOUT).await
}

/// Shared close implementation with an explicit grace timeout (tests
/// use a short one; `close` uses the real 5s constant).
///
/// The map lock is taken ONLY to remove the handle, and released before
/// the graceful close runs (the `spawn_blocking` await). Holding it
/// across the close would block every other command (send_prompt,
/// create_session, resume, ...) for up to the timeout on every session
/// close — a hung alayacore would freeze the whole backend.
pub async fn close_with_timeout(
    session_id: &str,
    sessions: &SessionMap,
    timeout: std::time::Duration,
) -> Result<(), String> {
    let handle = {
        let mut map = sessions.0.lock().await;
        map.remove(session_id)
    };
    let Some(handle) = handle else {
        return Err("Session not found".to_string());
    };
    let child_opt = handle.child.lock().unwrap().take();
    let stdin = handle.stdin.clone();
    let _ = tokio::task::spawn_blocking(move || {
        if let Some(mut child) = child_opt {
            alayacore::close_child_gracefully_with_timeout(&mut child, &stdin, timeout);
        }
    })
    .await;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    // M5 (D6): the pending-commands registry must stay BOUNDED — a CO
    // reply that never arrives (protocol anomaly / killed core) must
    // not grow the map forever. Insert evicts the OLDEST entries beyond
    // the cap (FIFO), so the newest calls — the ones most likely to
    // still get a reply — always survive. Mirrors the Go
    // pending_cmds_test.go cases.

    #[tokio::test]
    async fn pending_commands_bounded_by_cap() {
        let pc = PendingCommands::new();
        for i in 1..=(MAX_PENDING_COMMANDS + 100) {
            pc.insert(format!("call-{i}"), "cmd".to_string()).await;
        }
        assert_eq!(pc.len().await, MAX_PENDING_COMMANDS);
        // The 100 OLDEST entries were evicted; the newest survive.
        assert!(!pc.contains("call-1").await, "oldest entry survived the cap eviction");
        assert!(!pc.contains("call-100").await, "entry at the eviction boundary survived");
        assert!(pc.contains("call-101").await, "first surviving entry was evicted");
        assert!(pc.contains(&format!("call-{}", MAX_PENDING_COMMANDS + 100)).await, "newest entry was evicted");
    }

    #[tokio::test]
    async fn pending_commands_async_remove() {
        let pc = PendingCommands::new();
        pc.insert("call-1".to_string(), "model_set".to_string()).await;
        assert_eq!(pc.remove("call-1").await.as_deref(), Some("model_set"));
        assert!(pc.remove("call-1").await.is_none());
        assert_eq!(pc.len().await, 0);
    }

    #[test]
    fn pending_commands_blocking_remove_works_off_runtime() {
        // blocking_remove is the stdout-reader path: the reader runs on
        // a plain std thread with NO tokio runtime — blocking_lock must
        // work there (it panics when called inside a runtime).
        let pc = PendingCommands::new();
        let rt = tokio::runtime::Runtime::new().unwrap();
        rt.block_on(async {
            pc.insert("call-2".to_string(), "read_file".to_string()).await;
        });
        drop(rt);
        assert_eq!(pc.blocking_remove("call-2"), Some("read_file".to_string()));
        assert!(pc.blocking_remove("call-2").is_none());
    }

    #[tokio::test]
    async fn pending_commands_reinsert_does_not_duplicate_order() {
        let pc = PendingCommands::new();
        pc.insert("call-1".to_string(), "a".to_string()).await;
        pc.insert("call-1".to_string(), "a2".to_string()).await; // same id, new name
        for i in 1..=MAX_PENDING_COMMANDS {
            pc.insert(format!("call-{i}"), "cmd".to_string()).await;
        }
        // A duplicated order queue would evict two entries instead of one.
        assert_eq!(pc.len().await, MAX_PENDING_COMMANDS);
    }
    use std::process::{Command, Stdio};
    use std::time::Duration;

    /// Build a SessionMap with one handle backed by a real child process.
    fn map_with_session(
        mut child: std::process::Child,
        sid: &str,
    ) -> (SessionMap, Arc<tokio::sync::Mutex<Option<std::process::ChildStdin>>>) {
        let stdin = Arc::new(tokio::sync::Mutex::new(Some(child.stdin.take().unwrap())));
        let map = SessionMap(Arc::new(tokio::sync::Mutex::new(HashMap::new())));
        {
            let mut inner = map.0.try_lock().unwrap();
            inner.insert(
                sid.to_string(),
                SessionHandle {
                    stdin: stdin.clone(),
                    connected: Arc::new(AtomicBool::new(true)),
                    pending_commands: PendingCommands::new(),
                    child: Arc::new(std::sync::Mutex::new(Some(child))),
                    session_dir: PathBuf::from("/tmp/alayaface-test-session"),
                    owner: String::new(),
                },
            );
        }
        (map, stdin)
    }

    #[tokio::test]
    async fn close_releases_map_lock_while_graceful_close_runs() {
        // A child that ignores EOF: the graceful close must wait out its
        // timeout, which gives the test a deterministic window to observe
        // whether the SessionMap lock is still held.
        let child = Command::new("sleep")
            .arg("60")
            .stdin(Stdio::piped())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
            .expect("spawn sleep");
        let (map, _stdin) = map_with_session(child, "s1");

        let map_arc = map.0.clone();
        let closer = tokio::spawn(async move {
            close_with_timeout("s1", &SessionMap(map_arc), Duration::from_millis(600)).await
        });

        // Give the closer time to remove the handle and enter the
        // blocking graceful close.
        tokio::time::sleep(Duration::from_millis(150)).await;
        assert!(
            map.0.try_lock().is_ok(),
            "SessionMap lock must NOT be held while a session is closing \
             (graceful close can take seconds; holding it would freeze \
             every other command)"
        );

        closer.await.expect("close task panicked").expect("close should succeed");
    }

    #[tokio::test]
    async fn close_unknown_session_returns_not_found() {
        let map = SessionMap(Arc::new(tokio::sync::Mutex::new(HashMap::new())));
        let err = close("nope", &map).await.unwrap_err();
        assert_eq!(err, "Session not found");
    }
}

/// Helper: get a session handle by ID.
pub fn get<'a>(
    map: &'a HashMap<String, SessionHandle>,
    session_id: &str,
) -> Result<&'a SessionHandle, String> {
    map.get(session_id).ok_or_else(|| "Session not found".to_string())
}

File: src-go/internal/session/session.go
// Package session manages AlayaCore sessions.
//
// Each session corresponds to one `alayacore --rawio` subprocess.
// Port of src-tauri/src/session.rs.
package session

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"os/exec"
	"sync"
	"sync/atomic"
	"time"

	"github.com/google/uuid"

	"alayaface/src-go/internal/core"
	"alayaface/src-go/internal/hub"
	"alayaface/src-go/internal/tlv"
)

// ErrNotFound is returned when a session id is not in the manager.
// Message matches the Rust backend exactly ("Session not found").
var ErrNotFound = errors.New("Session not found")

// Session is a running alayacore session handle.
type Session struct {
	ID         string
	Stdin      io.WriteCloser
	Stdout     io.ReadCloser // owned by the reader goroutine
	Child      *exec.Cmd
	SessionDir string
	// Owner is the client identity that created/resumed this session
	// (empty = legacy/unknown). close_all_sessions reclaims only the
	// caller's own sessions, so one browser tab's page load never kills
	// another tab's live sessions (the Go backend is reachable from
	// multiple clients over LAN/SSH).
	Owner string

	stdinMu   sync.Mutex
	connected atomic.Bool
	// killOnce guarantees the child is killed/reaped exactly once even
	// when Close and the reader goroutine's EOF path race.
	killOnce sync.Once
	// PendingCmds maps call ID → command name (CI sent, CO not yet
	// received). The reader attaches the command name to CO frames,
	// since CO carries only the call ID. Bounded (FIFO eviction beyond
	// maxPendingCmds) so an unanswered CO cannot leak forever.
	PendingCmds *pendingCmds
}

// kill reaps the child process exactly once (idempotent).
func (s *Session) kill() {
	s.killOnce.Do(func() { core.KillChild(s.Child) })
}

// Connected reports whether the session's stdout pipe is still open.
func (s *Session) Connected() bool { return s.connected.Load() }

// setConnected stores the connection flag.
func (s *Session) setConnected(v bool) { s.connected.Store(v) }

// Manager is a shared map of session id → Session.
type Manager struct {
	mu       sync.Mutex
	sessions map[string]*Session
}

// NewManager creates an empty session manager.
func NewManager() *Manager {
	return &Manager{sessions: make(map[string]*Session)}
}

// Get returns a session by id.
func (m *Manager) Get(id string) (*Session, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	s, ok := m.sessions[id]
	if !ok {
		return nil, ErrNotFound
	}
	return s, nil
}

// Insert registers a session.
func (m *Manager) Insert(s *Session) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.sessions[s.ID] = s
}

// Remove deletes a session from the map and returns it.
func (m *Manager) Remove(id string) (*Session, bool) {
	m.mu.Lock()
	defer m.mu.Unlock()
	s, ok := m.sessions[id]
	if ok {
		delete(m.sessions, id)
	}
	return s, ok
}

// Len returns the number of active sessions.
func (m *Manager) Len() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.sessions)
}

// ForEach iterates over a snapshot of sessions, stopping early if fn
// returns true.
func (m *Manager) ForEach(fn func(*Session) bool) {
	m.mu.Lock()
	items := make([]*Session, 0, len(m.sessions))
	for _, s := range m.sessions {
		items = append(items, s)
	}
	m.mu.Unlock()
	for _, s := range items {
		if fn(s) {
			return
		}
	}
}

// takeAll removes every session from the map and returns them. The
// caller owns their teardown.
func (m *Manager) takeAll() []*Session {
	m.mu.Lock()
	defer m.mu.Unlock()
	items := make([]*Session, 0, len(m.sessions))
	for _, s := range m.sessions {
		items = append(items, s)
	}
	m.sessions = make(map[string]*Session)
	return items
}

// CloseAll kills every session (used on server shutdown). Runs the
// per-session teardown in parallel so N sessions cost one grace period,
// not N × grace period.
func (m *Manager) CloseAll() {
	items := m.takeAll()
	var wg sync.WaitGroup
	for _, s := range items {
		wg.Add(1)
		go func(s *Session) {
			defer wg.Done()
			s.kill()
		}(s)
	}
	wg.Wait()
}

// CloseAllGracefully closes every active session with the same
// cancel-first graceful sequence as Close (cancel → save → EOF → grace
// → SIGKILL). The frontend calls this once on page load so sessions
// orphaned by a page refresh (their windows are gone but the backend
// still holds the handles) are reclaimed — otherwise resume_session
// keeps failing with "Session is already active" until the backend is
// restarted. History is preserved up to each session's cancel point.
// Runs the per-session sequence in parallel (bounded by one grace
// period regardless of session count).
func (m *Manager) CloseAllGracefully() {
	m.CloseAllGracefullyFor("")
}

// CloseAllGracefullyFor closes the sessions owned by `owner` (empty =
// all). The frontend passes its per-page clientId so a page refresh
// reclaims only ITS orphaned sessions — with the Go backend reachable
// from multiple clients (LAN/SSH port forwarding), an unconditional
// close-all on every page load would kill the other clients' live
// sessions.
func (m *Manager) CloseAllGracefullyFor(owner string) {
	var items []*Session
	m.mu.Lock()
	for _, s := range m.sessions {
		if owner == "" || s.Owner == owner {
			items = append(items, s)
			delete(m.sessions, s.ID)
		}
	}
	m.mu.Unlock()
	var wg sync.WaitGroup
	for _, s := range items {
		wg.Add(1)
		go func(s *Session) {
			defer wg.Done()
			s.closeGracefully()
			log.Printf("[session] closed %s", s.ID)
		}(s)
	}
	wg.Wait()
}

// CreateConfig configures a new session.
type CreateConfig struct {
	ID          string // also names the session dir
	Binary      string
	ConfigPath  string
	SessionFile string
	SessionDir  string
	ToolConfirm string
	// BuiltinTools: nil = don't pass --builtin-tools (all tools);
	// pointer to "" = explicitly no builtin tools (Plan Sessions);
	// pointer to a list = those tools only.
	BuiltinTools *string
	SystemPrompt string
	// ReasoningLevel is the session's initial reasoning level (0|1|2),
	// passed to alayacore as --reasoning-level=<n>.
	ReasoningLevel int
	WorkDir        string // child cwd (per-plan isolation; "" = backend cwd)
	// Owner is the client identity that created the session (empty =
	// legacy). Used by close_all_sessions to reclaim only one client's
	// orphaned sessions.
	Owner string
}

// Create spawns alayacore, registers the session, and starts the stdout
// reader. The session is fully registered and the reader running before
// Create returns (same ordering as Rust), so onStatus cannot arrive
// before onSessionCreated on the client.
func (m *Manager) Create(cfg CreateConfig, h *hub.Hub, cache *ModelCache) (*Session, error) {
	proc, err := core.Spawn(cfg.Binary, cfg.ConfigPath, cfg.SessionFile, cfg.ToolConfirm, cfg.BuiltinTools, cfg.SystemPrompt, cfg.ReasoningLevel, cfg.WorkDir)
	if err != nil {
		return nil, fmt.Errorf("Failed to start alayacore: %w", err)
	}

	s := &Session{
		ID:          cfg.ID,
		Stdin:       proc.Stdin,
		Stdout:      proc.Stdout,
		Child:       proc.Cmd,
		SessionDir:  cfg.SessionDir,
		Owner:       cfg.Owner,
		PendingCmds: newPendingCmds(),
	}
	s.setConnected(true)

	m.Insert(s)

	// Broadcast connected:true BEFORE starting the stdout reader. If the
	// child dies instantly (bad args, missing binary, ...), the reader
	// immediately observes EOF and broadcasts connected:false — with the
	// old ordering (reader first) the false could arrive before the true,
	// leaving the client believing a dead session is connected.
	h.Broadcast(hub.NewEvent("core-status", StatusEvent{
		SessionID: s.ID,
		Connected: true,
		Message:   fmt.Sprintf("Connected to alayacore (%s)", cfg.Binary),
	}))

	s.startReader(h, cache)
	log.Printf("[session] created %s (dir %s)", s.ID, cfg.SessionDir)
	return s, nil
}

// gracefulCloseTimeout is how long close_session waits for alayacore to
// exit after the cancel-first sequence (cancel → save → EOF) before
// SIGKILLing it. Mirrors alayacore::GRACEFUL_CLOSE_TIMEOUT (Rust).
const gracefulCloseTimeout = 5 * time.Second

// Close removes the session from the map and closes it cancel-first:
// CI "cancel" (aborts the running task, auto-saving up to the cancel
// point) → CI "save" → close stdin (EOF → exits immediately) → SIGKILL
// only after the grace period.
func (m *Manager) Close(id string) error {
	s, ok := m.Remove(id)
	if !ok {
		return ErrNotFound
	}
	s.closeGracefully()
	log.Printf("[session] closed %s", id)
	return nil
}

// closeGracefully runs the cancel-first shutdown sequence for one session.
//
//  1. CI `cancel` — aborts the running task (alayacore cancels its
//     per-task context; the task completes through handleTaskDone, which
//     AUTO-SAVES the conversation up to the cancel point). Best-effort:
//     nothing to cancel / dead child produce an error that is ignored.
//  2. CI `save` — persist the conversation (best-effort).
//  3. EOF: close the stdin pipe. With the task canceled (or none),
//     alayacore exits immediately — it no longer drains a long-running
//     task to completion (cancel-first means Stop/closing a window
//     never waits for a task to finish, and history is kept up to the
//     cancel point instead of being lost to SIGKILL).
//  4. Wait for the natural exit, SIGKILL after the grace period.
//
// It does NOT call cmd.Wait() itself: reaping is owned by the reader's
// disconnect path (killOnce → core.KillChild), which fires exactly when
// the child's stdout closes — i.e. when the child exits on its own. So
// we only poll Connected() to learn the child has gone. Calling Wait
// here would race with that path (os/exec forbids concurrent Waits).
func (s *Session) closeGracefully() {
	// The whole cancel → save → EOF sequence runs under the stdin lock
	// so a concurrent SendPrompt/WriteFrames can neither interleave
	// between the CI frames nor be mid-write when the pipe closes
	// (which would leave a partial TLV frame on the wire).
	s.stdinMu.Lock()
	// 1. Cancel the active task first (fire-and-forget — no CO wait;
	//    errors ignored).
	_, _ = s.sendCmdLocked("cancel", "")
	// 2. Ask alayacore to persist the conversation (best-effort: a dead
	//    child produces a write error, which is ignored).
	_, _ = s.sendCmdLocked("save", "")
	// 3. EOF: close the stdin pipe. With no active task (canceled), the
	//    child exits immediately.
	_ = s.Stdin.Close()
	s.stdinMu.Unlock()
	// 4. Wait for the reader to observe the natural exit (stdout EOF →
	//    disconnect → killOnce reaps the child).
	deadline := time.Now().Add(gracefulCloseTimeout)
	for s.Connected() {
		if time.Now().After(deadline) {
			// 5. Fallback: SIGKILL (killOnce also reaps; idempotent if
			//    the reader already killed).
			s.kill()
			return
		}
		time.Sleep(50 * time.Millisecond)
	}
}

// WriteFrame sends a raw TLV frame to the session's stdin.
func (s *Session) WriteFrame(tag, value string) error {
	return s.WriteFrames([]tlv.Frame{{Tag: tag, Value: value}})
}

// WriteFrames sends a batch of TLV frames under a single stdin lock.
// Used for multi-frame messages (e.g. media + text + UE flush) so a
// concurrent CI command cannot interleave between frames — mirrors the
// Rust send_prompt which holds the stdin lock for the whole sequence.
func (s *Session) WriteFrames(frames []tlv.Frame) error {
	if !s.Connected() {
		return errors.New("Session is disconnected")
	}
	s.stdinMu.Lock()
	defer s.stdinMu.Unlock()
	for _, f := range frames {
		if err := tlv.WriteFrame(s.Stdin, f.Tag, f.Value); err != nil {
			return fmt.Errorf("write error: %w", err)
		}
		preview := f.Value
		if len(preview) > 200 {
			preview = preview[:200]
		}
		log.Printf("[tlv] >> %s %s %db %s", s.ID, f.Tag, len(f.Value), preview)
	}
	return nil
}

// SendCmd sends a CI (command input) frame and returns the generated call ID.
func (s *Session) SendCmd(name, input string) (string, error) {
	s.stdinMu.Lock()
	defer s.stdinMu.Unlock()
	return s.sendCmdLocked(name, input)
}

// sendCmdLocked sends a CI frame while stdinMu is held. The call ID →
// name mapping is registered BEFORE the frame is written — the CO reply
// can arrive as soon as the CI frame is flushed.
func (s *Session) sendCmdLocked(name, input string) (string, error) {
	id := uuid.NewString()
	s.PendingCmds.Store(id, name)
	payload, err := json.Marshal(tlv.CmdMsg{ID: id, Name: name, Input: input})
	if err != nil {
		return "", err
	}
	if err := tlv.WriteFrame(s.Stdin, tlv.TagCmdInput, string(payload)); err != nil {
		s.PendingCmds.Delete(id)
		return "", err
	}
	preview := payload
	if len(preview) > 200 {
		preview = preview[:200]
	}
	log.Printf("[tlv] >> %s %s %db %s", s.ID, tlv.TagCmdInput, len(payload), preview)
	return id, nil
}
